### PR TITLE
fix: escape paths with spaces in pybind11-config

### PR DESCRIPTION
**Pull request by MarkusBauer**
## Description

tl;dr: `pybind11 --includes` returns invalid arguments when the project's directory contains spaces.

When you build an extension manually, the [docs](https://pybind11.readthedocs.io/en/stable/compiling.html#building-manually) suggest this command: `c++ ... $(python3 -m pybind11 --includes) ...`. 
However, this command fails when the project's directory contains a space in its path. 
The command would then expand to `c++ ... -I/home/user/a b/c -I/python ...`, where `/home/user/a` is a in invalid include directory, and `b/c` is an invalid compiler argument.

Even if no manual build is desired this bug might affect users: for example, we're currently building `clang-tidy` in our project, which requires all include paths. 

To solve this issue, this PR escapes include paths of `pybind11 --includes` before printing. Paths without spaces/special characters are printed as-is, while paths with spaces are wrapped in `''`. Thus, the final command line of custom builds will be valid even if spaces occur: `c++ ... '-I/home/user/a b/c' -I/python ...`

### CODE REVIEW

File: pybind11/__main__.py [hunk 1]
```diff
@@ -21,7 +22,7 @@ def print_includes() -> None:
         if d and d not in unique_dirs:
             unique_dirs.append(d)
 
-    print(" ".join("-I" + d for d in unique_dirs))
+    print(" ".join(shlex.quote("-I" + d) for d in unique_dirs))
```
Comment (rwgk) [inline pybind11/__main__.py:25]: The `shlex.quote` documentation has this big red warning:

https://docs.python.org/3/library/shlex.html#shlex.quote

> Warning The shlex module is only designed for Unix shells.

Did you already think about this? What is the behavior under Windows? From really rusty memory, I think Windows may only understand `"` but not `'`, is that true?

Comment (MarkusBauer) [reply pybind11/__main__.py:25]: Just tested the behavior on Windows:
- everything works fine with Powershell
- old CMD does not understand the new quoting
- backslashes trigger the escaping, so any path with backslashes will get invalid on Windows

I see three possible solutions, both have their drawbacks:
- pull in an additional dependency ([oslex](https://pypi.org/project/oslex/) or [mslex](https://pypi.org/project/mslex/))
- do the quoting manually on Windows (which might or might not be a complete implementation - at least we're not talking about untrusted data here)
- disable the quoting on Windows (and accept the bug on one platform)

I can update the implementation if you chose your preferred way.

Comment (rwgk) [reply pybind11/__main__.py:25]: oof :-)

I'm really out of touch with the Windows world.

> pull in an additional dependency (oslex or mslex)

That sounds best to me.

But not sure what makes sense:

* try-except import oslex or mslex
* if available: all good
* if not available: check for obvious failure conditions (e.g. spaces, maybe backslashes?) and raise an exception
* otherwise just skip the quoting and hope for the best

Comment (MarkusBauer) [reply pybind11/__main__.py:25]: What about such a construct? Pick `oslex` if available, `shlex` as linux fallback and stupid windows fallback otherwise? The Windows fallback catches at least the most common case (spaces).

```python
try:
    from oslex import quote
except:
    import os
    if os.name != 'nt':
        from shlex import quote
    else:
        def quote(s):
            return '"' + s + '"' if ' ' in s else s
```

Comment (rwgk) [reply pybind11/__main__.py:25]: Looks good to me!

Nits:

`except ImportError:`

```
if os.name != "nt":
```
(I think black will insist on that)

Maybe also add a one-line comment for the fallback `quote()` implementation, something like

```
# Minimal attempt, handling only the most common case.
```

Comment (MarkusBauer) [reply pybind11/__main__.py]: I added oslex and the proposed windows fallback. 

However, it looks to me like pybind11 has zero python runtime dependencies so far, and therefore does not have anything like `requirements.txt`. Did I miss something?
If this is true, I suggest that we don't introduce that and keep `oslex` purely optional.

Comment (rwgk) [reply pybind11/__main__.py]: The current version of this PR looks good to me.

We have requirements.txt only under `tests/`.

> If this is true, I suggest that we don't introduce that and keep oslex purely optional.

"that" == requirements.txt?

Sounds good to me: this PR only makes things better, without introducing hard new dependencies.

Comment (MarkusBauer) [reply pybind11/__main__.py]: `"that" == requirements.txt?` - yes. Not necessary to introduce it here.

**Review by rwgk — APPROVED** [review]
Thanks!

Waiting for @henryiii to help with a 2nd set of eyes.

File: pybind11/__main__.py (hunk 1 above)
Comment (henryiii) [reply pybind11/__main__.py]: Little worried about adding a dependency with two stars on GitHub. :) Also just assuming it might be present, rather than making it, for example, conditional. I think this is a bit better than the status quo, though, so I'd be fine with it.

File: pybind11/__main__.py [hunk 2]
```diff
@@ -21,7 +33,7 @@ def print_includes() -> None:
         if d and d not in unique_dirs:
             unique_dirs.append(d)
 
-    print(" ".join("-I" + d for d in unique_dirs))
+    print(" ".join(quote("-I" + d) for d in unique_dirs))
```
Comment (henryiii) [inline pybind11/__main__.py:36]:
```suggestion
    print(" ".join(quote(f"-I{d}") for d in unique_dirs))
```

File: pybind11/__main__.py [hunk 3]
```diff
@@ -7,6 +7,18 @@
...
+    from oslex import quote
+except ImportError:
+    import os
+
+    if os.name != "nt":
+        from shlex import quote
+    else:
+        # minimal attempt, handling only the most common case (spaces in path)
+        def quote(s: str) -> str:
+            return '"' + s + '"' if " " in s else s
```
Comment (henryiii) [inline pybind11/__main__.py:20]:
```suggestion
            return f'"{s}"' if " " in s else s
```

**Comment by henryiii** [comment]
I'd forgotten about this. Unless mslex modernizes a bit (it currently uses setup.py with deprecated arguments, like test_requires), I think I'd like to just have the custom form (from meson). I'll update this.

File: pybind11/__main__.py [hunk 4]
```diff
@@ -2,12 +2,35 @@
...
+    # https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/
+    UNSAFE = re.compile("[ \t\n\r]")
+
+    def quote(s: str) -> str:
+        if s and not UNSAFE.search(s):
+            return s
+
+        # Paths cannot contain a '"' on Windows, so we don't need to worry
+        # about nuanced counting here.
+        return f'"{s}"'
```
Comment (QuLogic) [inline pybind11/__main__.py:28]: Does this handle a trailing backslash correctly? I see all the internally-generated paths use `os.path.join`, and are probably safe, but I'm not 100% sure about `sysconfig.get_path`.

Comment (MarkusBauer) [reply pybind11/__main__.py:28]: It doesn't. We could use `if s.endswith("\\") s += "\\"` in line 28 to fix this. 
Cmd does not accept escaped `\\` in general, but in front of a terminal `"` it's ok.

Comment (henryiii) [reply pybind11/__main__.py:28]:
```suggestion
        # Paths cannot contain a '"' on Windows, so we don't need to worry
        # about nuanced counting here.
        return f'"{s}\\"' if s.endswith("\\") else f'"{s}"'
```
?

Comment (QuLogic) [reply pybind11/__main__.py:28]: Unless you want to support something like `c:\` as an include path, maybe stripping trailing backslashes would be safe enough?

**Review by rwgk — APPROVED** [review]
... alright ...

I cannot honestly write "looks good to me" here, but it seem beyond anyone's capacity (definitely my own) to do something actually good here :-)

I'm holding out hope that some future Python version will come with a platform-aware rock-solid `quote()` implementation out of the box.

**Comment by QuLogic** [comment]
FWIW, I just tried this on Windows, and everything is quoted enough to work properly, and `cl` no longer warns about non-existent paths.

**Comment by henryiii** [comment]
Released.
